### PR TITLE
stats: add/test heterogenous set of StatNameStorage objects.

### PR DESCRIPTION
**Pull request by jmarantz**
Description: This PR breaks out one of the concepts from #6161 -- an efficient mechanism for remembering a set of StatNames to reject, to use for caching the results of the stats-matcher.
Risk Level: low -- not used yet.
Testing: //test/...
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

File: source/common/stats/symbol_table_impl.cc [hunk 1]
```diff
@@ -317,6 +317,36 @@ void StatNameStorage::free(SymbolTable& table) {
...
+  //
+  // There's a performance risk here, if removing elements via
+  // flat_hash_set::begin() is inefficient to use in a loop like this. One can
+  // imagine a hash-table implementation where the performance of this
+  // usage-model would be poor. However, tests with 100k elements appeared to
+  // run quickly when compiled for optimization, so at present this is not a
+  // performance issue.
+
+  while (!empty()) {
+    auto storage = extract(begin());
```
Comment (ambuc) [inline source/common/stats/symbol_table_impl.cc]: Is there a way to fully-qualify this as a function within flat_hash_map? Just as an aesthetic thing.

Comment (ambuc) [inline source/common/stats/symbol_table_impl.cc]: Is there a way to fully-qualify `extract()` as a method on flat_hash_map? Just an aesthetic thing.

File: source/common/stats/symbol_table_impl.h [hunk 2]
```diff
@@ -132,13 +132,15 @@ class SymbolTableImpl : public SymbolTable {
...
   void incRefCount(const StatName& stat_name) override;
-  SymbolTable::StoragePtr join(const std::vector<StatName>& stat_names) const override;
+  StoragePtr join(const std::vector<StatName>& stat_names) const override;
   void populateList(const absl::string_view* names, uint32_t num_names,
                     StatNameList& list) override;
   StoragePtr encode(absl::string_view name) override;
   void callWithStringView(StatName stat_name,
                           const std::function<void(absl::string_view)>& fn) const override;
 
+  void encode(absl::string_view name, Encoding& encoding);
```
Comment (ambuc) [inline source/common/stats/symbol_table_impl.h]: Is this implemented?

File: test/common/stats/symbol_table_impl_test.cc [hunk 3]
```diff
@@ -507,6 +507,20 @@ TEST_P(StatNameTest, MutexContentionOnExistingSymbols) {
   }
 }
 
+TEST_P(StatNameTest, SharedStatNameStorageSet) {
```
Comment (ambuc) [inline test/common/stats/symbol_table_impl_test.cc]: nit: Rename this to describe the aspect of the sharedstatenamestorageset we're testing.

File: test/common/stats/symbol_table_impl_test.cc [hunk 4]
```diff
@@ -507,6 +507,20 @@ TEST_P(StatNameTest, MutexContentionOnExistingSymbols) {
...
+  {
+    for (int i = 0; i < 10; ++i) {
+      std::string foo = absl::StrCat("foo", i);
+      auto insertion = set.insert(StatNameStorage(foo, *table_));
+      StatNameTempStorage temp_foo(foo, *table_);
+      auto found = set.find(temp_foo.statName());
+      EXPECT_EQ(found->statName().data(), insertion.first->statName().data());
+    }
+  }
+  set.free(*table_);
```
Comment (ambuc) [inline test/common/stats/symbol_table_impl_test.cc]: Worth testing an assert failure if free isn't called?

File: source/common/stats/symbol_table_impl.cc (hunk 1 above)
Comment (jmarantz) [reply source/common/stats/symbol_table_impl.cc]: I was on the fence anyway about inheriting vs containing the hash-set. I switched to containment and now I think this is much clearer.

File: source/common/stats/symbol_table_impl.h (hunk 2 above)
Comment (jmarantz) [reply source/common/stats/symbol_table_impl.h]: no it was detritus; deleted. good catch.

File: test/common/stats/symbol_table_impl_test.cc (hunk 3 above)
Comment (jmarantz) [reply test/common/stats/symbol_table_impl_test.cc]: done

File: test/common/stats/symbol_table_impl_test.cc (hunk 4 above)
Comment (jmarantz) [reply test/common/stats/symbol_table_impl_test.cc]: done

File: test/common/stats/symbol_table_impl_test.cc [hunk 5]
```diff
@@ -507,6 +507,44 @@ TEST_P(StatNameTest, MutexContentionOnExistingSymbols) {
...
+  EXPECT_EQ(1, set1.size());
+  EXPECT_EQ(0, set2.size());
+  set1.swap(set2);
+  EXPECT_EQ(0, set1.size());
+  EXPECT_EQ(1, set2.size());
+  set2.free(*table_);
+}
+
+TEST_P(StatNameTest, SharedStatNameStorageSetIfDestroyedWithoutFree) {
+#ifndef NDEBUG
```
Comment (htuch) [inline test/common/stats/symbol_table_impl_test.cc]: I don't think tests for ASSERTs make sense; these are for debug and invariants as code comments.

Comment (jmarantz) [reply test/common/stats/symbol_table_impl_test.cc]: OK; @ambuc had asked for it, but I don't feel that strongly either way. Removed.

**Pull request merged**